src/main/java/online/talkandtravel/service/impl/OnlineServiceImpl.java

```diff
@@ -0,0 +1,140 @@
+package online.talkandtravel.service.impl;
+
+import static online.talkandtravel.util.AuthenticationUtils.getUserFromPrincipal;
+
+import java.security.Principal;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import online.talkandtravel.model.dto.user.UserOnlineStatusDto;
+import online.talkandtravel.model.entity.User;
+import online.talkandtravel.service.OnlineService;
+import online.talkandtravel.util.RedisUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.data.redis.core.RedisTemplate;
+import org.springframework.messaging.simp.SimpMessagingTemplate;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+@Log4j2
+public class OnlineServiceImpl implements OnlineService {
+
+    private final RedisTemplate<String, String> redisTemplate;
+    private final SimpMessagingTemplate messagingTemplate;
+    public static final String USER_STATUS_KEY = "user:%s:isOnline";
+    public static final String USER_STATUS_KEY_PATTERN = "user:*:isOnline";
+
+    public static final String USERS_ONLINE_STATUS_ENDPOINT = "/users/onlineStatus";
+
+    @Value("${USER_ONLINE_STATUS_EXPIRATION_DURATION_IN_MIN}")
+    public Long KEY_EXPIRATION_DURATION_IN_MIN;
+
+    @Override
+    public void updateUserOnlineStatus(Principal principal, Boolean isOnline) {
+        User user = getUserFromPrincipal(principal);
+        log.info("updateUserOnlineStatusById userId: {}, isOnline: {}", user.getId(), isOnline);
+        String key = getRedisKey(user.getId());
+
+        if (isOnline) {
+            Duration keyDuration = Duration.ofMinutes(KEY_EXPIRATION_DURATION_IN_MIN);
+            redisTemplate.opsForValue().set(key, isOnline.toString(), keyDuration);
+        } else {
+            redisTemplate.delete(key);
+        }
+
+        notifySubscribedUsers(user.getId(), isOnline);
+    }
+
+    @Override
+    public void notifyUserOnlineStatusUpdated(Long userId, Boolean isOnline) {
+        notifySubscribedUsers(userId, isOnline);
+    }
+
+    /**
+     * Provides an initial snapshot of all user's online statuses
+     *
+     * @return a map where the key is the user ID and the value is a boolean indicating
+     * whether the user is online (true) or offline (false)
+     */
+    @Override
+    public Map<Long, Boolean> getAllUsersOnlineStatuses() {
+        log.info("getAllUsersOnlineStatuses");
+
+        Set<String> keys = redisTemplate.keys(USER_STATUS_KEY_PATTERN);
+        List<Long> userIdList = getUserIdFromKeys(keys);
```

Comment:
add here check 'keys' for null value and if it null - what will happen? (handle case where 'keys' is null)